src/emulate.c

```diff
@@ -377,9 +447,10 @@ enum {
 };
 
 #if RV32_HAS(GDBSTUB)
-#define RVOP_NO_NEXT(ir) (!ir->next | rv->debug_mode)
+#define RVOP_NO_NEXT(ir) \
+    (!ir->next | rv->debug_mode IIF(RV32_HAS(SYSTEM))(| rv->is_trapped, ))
 #else
-#define RVOP_NO_NEXT(ir) (!ir->next)
+#define RVOP_NO_NEXT(ir) (!ir->next IIF(RV32_HAS(SYSTEM))(| rv->is_trapped, ))
```

Comment:
Should it be `(!ir->next IIF(RV32_HAS(SYSTEM))(|| rv->is_trapped, ))` instead?

Reply:
Single `|` has same effect than double `||` since the all condition are boolean.

Reply:
The current semantics equivalent, but I'm worrying about the support of nested interrupt handling in the future. The `is-trapped` might become a integer as a counter.

Reply:
> The current semantics is equivalent, but I'm worrying about the support of nested interrupt handling in the future. The `is-trapped` might become a integer as a counter.

If used as a counter, the implementation of `sret` should decrement `is_trapped` rather than simply setting it to false? Also, other places using `is_trapped` should increment it?

However, I'm curious in the current emulation, have you observed nested traps occurring, and do these changes need to be made to handle them?

Reply:
> However, I'm curious in the current emulation, have you observed nested traps occurring, and do these changes need to be made to handle them?

In current, no (if I'm not mistaken). But we need it if we want to handle the page fault during handling the external interrupt, right?

Reply:
> > However, I'm curious in the current emulation, have you observed nested traps occurring, and do these changes need to be made to handle them?
> 
> In current, no (if I'm not mistaken). But we need it if we want to handle the page fault during handling the external interrupt, right?

Indeed. However, we currently don't have a test suite for this situation. The best test suite would be the Linux kernel itself, and we might need an asynchronous mechanism (e.g., a signal thread) to periodically intercept the external interrupt. 

For simplicity, I would prefer to keep track of this issue and address it only when it becomes necessary.